### PR TITLE
Fix #610 and  Improvements

### LINKED ISSUE #610: Problem with canonical link
If the user fills in the Canonical Link with any text that is not a valid URL, we encounter an error similar to the one found below.

https://github.com/cnvs/canvas/blob/4aeaa2eb5c2b421ddecb878a64cd0c160413fc17/resources/stubs/views/blog/show.stub#L82-L90

image

### PR DESCRIPTION
**Pull request by albertoelias123**
Fixes issue #610 in the commit https://github.com/cnvs/canvas/commit/3e47fea1b35c35b78812010da3d060540c15bdec
I believe it should be discussed about issue # 610, because the problem actually occurs when the user does not put a valid link (http ...). The correct thing would be not to allow an invalid URL to be registered? Should this be checked on the front, on the model or on the controller?

I removed the Factory from User, because in the latest version of Laravel a Factory of User is already included. If the user has a different model from the standard he is free to modify it.

Updated the CONTRIBUTING file

Do you think it is necessary to explain more about the --data on canvas: setup option?
Do you find it interesting to talk about factories? Should they be published?

I intend to send other PR's, did you find this one? Suggest any modifications?

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -62,7 +64,7 @@ From your Laravel app, create the authentication system and run the following co
 composer require laravel/ui
 
 php artisan ui vue --auth
-php artisan:migrate
+php artisan migrate
```
Comment (austintoddj) [inline .github/CONTRIBUTING.md]: This change is fine, good catch.

File: readme.md [hunk 2]
```diff
@@ -205,6 +210,11 @@ Finally, just create `index.blade.php` and `show.blade.php` inside a `/views/blo
 
 Just run `php artisan canvas:setup` after installing Canvas. Then, navigate your browser to `http://your-app.test/blog` or any other URL that is assigned to your application. This command scaffolds a default frontend for your entire blog!
 
+If you want test data from the application, use the `--data` option:
+```bash
+php artisan canvas:setup --data
+```
+
```
Comment (austintoddj) [inline readme.md]: The `--data` option was removed from the project, so this can go.

File: src/CanvasServiceProvider.php [hunk 3]
```diff
@@ -29,6 +29,7 @@ public function boot()
         $this->registerPublishing();
         $this->loadViewsFrom(__DIR__.'/../resources/views', 'canvas');
         $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'canvas');
+        $this->loadFactoriesFrom(__DIR__.'/Storage/factories');
```
Comment (austintoddj) [inline src/CanvasServiceProvider.php]: Since we're not using the seeders anymore we shouldn't need this.

File: src/Console/SetupCommand.php [hunk 4]
```diff
@@ -15,7 +15,7 @@ class SetupCommand extends Command
      *
      * @var string
      */
-    protected $signature = 'canvas:setup';
+    protected $signature = 'canvas:setup {--data : Specifies that demo data should be seeded}';
```
Comment (austintoddj) [inline src/Console/SetupCommand.php]: Same explanation as above.

File: src/Console/SetupCommand.php [hunk 5]
```diff
@@ -50,9 +50,54 @@ public function handle()
...
 
+        // Optionally seed the database with demo data
+        if ($this->option('data')) {
+            if (\App\User::find(1)) {
+                $this->seed();
+            } else {
+                $this->error('No users found. Please create a user and re-run the setup.');
+            }
+        }
+
```
Comment (austintoddj) [inline src/Console/SetupCommand.php]: Same explanation as above.

File: src/Console/SetupCommand.php [hunk 6]
```diff
@@ -50,9 +50,54 @@ public function handle()
...
+            $post->topic()->attach(
+                $topics->random(1)->pluck('id')->toArray()
+            );
+        });
+
+        // Seed the Canvas View data
+        factory(\Canvas\View::class, 2500)->create();
+    }
+
```
Comment (austintoddj) [inline src/Console/SetupCommand.php]: Same explanation as above.

File: src/Storage/factories/PostModelFactory.php [hunk 7]
```diff
@@ -12,7 +12,10 @@
         'featured_image'         => $faker->imageUrl(),
         'featured_image_caption' => $faker->sentence,
         'user_id'                => function () {
-            return factory(\Illuminate\Foundation\Auth\User::class)->create()->id;
+            $user = \App\User::inRandomOrder()->first();
+            if(is_null($user))
+                $user = factory(\App\User::class)->create();
+            return $user->id;
```
Comment (austintoddj) [inline src/Storage/factories/PostModelFactory.php]: The factory here is only utilized for the unit tests, so the line you removed should be sufficient.

File: src/Storage/factories/TagModelFactory.php [hunk 8]
```diff
@@ -7,7 +7,10 @@
         'slug'    => $faker->slug,
         'name'    => $faker->word,
         'user_id' => function () {
-            return factory(\Illuminate\Foundation\Auth\User::class)->create()->id;
+            $user =\App\User::inRandomOrder()->first();
+            if(is_null($user))
+                $user = factory(\App\User::class)->create();
+            return $user->id;
```
Comment (austintoddj) [inline src/Storage/factories/TagModelFactory.php]: Same comment as above.

File: src/Storage/factories/TopicModelFactory.php [hunk 9]
```diff
@@ -7,7 +7,10 @@
         'slug'    => $faker->slug,
         'name'    => $faker->word,
         'user_id' => function () {
-            return factory(\Illuminate\Foundation\Auth\User::class)->create()->id;
+            $user = \App\User::inRandomOrder()->first();
+            if(is_null($user))
+                $user = factory(\App\User::class)->create();
+            return $user->id;
```
Comment (austintoddj) [inline src/Storage/factories/TopicModelFactory.php]: Same comment as above.

File: src/Storage/factories/ViewModelFactory.php [hunk 10]
```diff
@@ -29,7 +29,10 @@
 
     return [
         'post_id' => function () {
-            return factory(\Canvas\Post::class)->create()->id;
+            $post = \Canvas\Post::inRandomOrder()->first();
+            if(is_null($post))
+                $post = factory(\Canvas\Post::class)->create();
+            return $post->id;
```
Comment (austintoddj) [inline src/Storage/factories/ViewModelFactory.php]: Same comment as above.

**Comment by austintoddj** [comment]
Thanks for submitting this @albertoelias123. I left some comments, and if you go through and make some adjustments we can merge it in, because some of the changes are necessary.

If you could make sure whitespace or indenting changes get removed that'd make it a lot easier to review the actual code. So the only 2 changes that I think need to be in this PR are:

- The documentation in the contributing guide that fixes `artisan migrate`
- The `parse_url()` method usage in the .stub file

Once I see those changes get added I'll re-review it, and thanks!

File: readme.md (hunk 2 above)
Comment (albertoelias123) [reply readme.md]: I implemented the --data option again to facilitate the development of the package and to allow users to quickly test the system. Do you believe that it is better not to implement this option?

File: src/CanvasServiceProvider.php (hunk 3 above)
Comment (albertoelias123) [reply src/CanvasServiceProvider.php]: I added the factories to allow the user to test the package easily. Do you believe it is better to remove this?

File: src/Storage/factories/PostModelFactory.php (hunk 7 above)
Comment (albertoelias123) [reply src/Storage/factories/PostModelFactory.php]: I don't understand, this factory is necessary for unit tests and therefore it should not be removed?

**Comment by albertoelias123** [comment]
I believe I made the requested corrections, @austintoddj could you take a look? I did not understand, the problem that occurred with Style (continuous-integration / styleci / pr)

**Comment by austintoddj** [comment]
Thanks for making those changes. If you could also revert the Table of Contents section in both the contributing file as well as the readme, I'd appreciate that. I just want to have the content listed that follows, not precedes.

Also, if you could change the base of this PR to merge into the `develop` branch instead of `master`:
Screen Shot 2020-01-28 at 5 37 20 AM

And don't worry about the StyleCI checks, those will be taken care of by myself prior to merging it into master.

**Pull request merged**